### ISSUE TITLE
Scrubber doesn't handle thorough tables

### ISSUE THREAD

**Issue by mswintermeyer** [palantir/atlasdb#3466]
The point of the scrubber is so that tables can be cleared of sensitive info before the sweeper gets around to them. However, it only operates in conservative mode. That means that sensitive row or column names are never deleted. Aka thorough tables don't benefit from scrubber and instead have to wait for sweep to complete.

**Comment by mswintermeyer** [comment]
Also FYI seems there are a few open, potentially related, issues regarding scrubber: #2408, #2266, #1697

**Comment by mswintermeyer** [comment]
In addition to more scrubber tests, it'd probably be useful to have tests for different sweep strategies that test how hard delete, scrubber, targeted sweep, and background sweep all interact

**Comment by mswintermeyer** [comment]
Some theorizing from the slack room: 

Scrubber is supposed to be very fast, i.e. targeted sweep just for hard deleted cells (before targeted sweep was a thing). As an example, aggressive hard delete blocks transactions finishing until scrubber scrubs them (I believe via calls scrubber’s scrubImmediately). Targeted sweep is weighed down by other cells and so isn’t fast enough.

Though now that targeted sweep exists, perhaps it’s easy to rewrite scrubber so it is actually targeted sweep just for hard deleted cells.

**Comment by sandorw** [comment]
Agreed that we can share more code between the scrubber and targeted sweep, likely by completely removing the scrub related code in favor of targeted sweep, as long as we satisfy all the requirements.

I haven't done a deep code dive, but I suspect that `HARD_DELETE` scrubbed transactions are pretty similar to current targeted sweep. The targeted sweep queue obviously contains more data, so if we combine these queues, we need to ensure that we're allocating enough threads/shards that scrub isn't falling unacceptably behind.

Currently, aggressive hard delete transactions:
- wait for the transaction to commit
- wait for the immutable timestamp to pass (polling every 500ms)
- deletes the value and applies a sentinel

This means that `AGGRESSIVE_HARD_DELETE` transactions are also very similar to targeted sweep, except that, again, we act more quickly if the targeted sweep queue is behind. We also potentially save a few seconds by calling sweep/scrub directly rather than waiting for a thread to drain the queue once the immutable timestamp has passed.

Questions:
- I suspect that the immutable timestamp is often O(minutes) to O(hours) behind for the big internal product. Is this a sufficient turnaround window on `AGGRESSIVE_HARD_DELETE` transactions? After that point, assuming that the targeted sweep queue isn't falling behind, is it acceptable to take ~a minute to remove the value after the immutable timestamp has passed?
- Do we have any requirement for how quickly `HARD_DELETE` transactions get cleaned up?
- Is there a distinction between how quickly we need to delete the value and evidence of the row/column keys?

Another option is to maintain separate targeted sweep queues for hard deleted transactions, which ensures that we won't fall behind if the queue is overly full. Sweep threads could give priority to these queues. If necessary, we could also allow on demand sweeps of these queues to emulate the current aggressive hard delete behavior.

If waiting for the immutable timestamp to pass is insufficient (O(hours) for the big internal product at times, by my understanding), then there are other more complicated options:
- Directly delete the cell in the KVS. This violates the transaction protocol, and without using sentinels, other transactions won't be aware that there was a value there that they should have read. This is the quickest way to actually remove the value and the associated row/column keys.
- Delete the value, applying a sentinel as though we were doing conservative sweep. Enqueue this delete in the targeted sweep queue for cleanup. This preserves the protocol (other transactions realize that they should have found a value and retry), but also leaves a sentinel at that row/column key (as well as a temporary entry in the targeted sweep queue) which indicates existence until the immutable timestamp moves past.

I don't like the first option at all. The second adds complexity but would delete the values faster than current aggressive hard delete transactions (but not the row/column keys).

**Comment by sandorw** [comment]
As for whether or not targeted sweep is fast enough (via current queues or additional queues for scrubbing), we should gather metrics on this. We already have metrics across many products to show how far behind the immutable timestamp sweep is that generally look very good, although it depends on the product and the write workflows. In typical cases, we've swept to within <30s of live writes, although for these products the transaction size is small and the immutable timestamp moves quickly.

**Comment by mswintermeyer** [comment]
In response to your questions:
1. My guess is waiting O(low-ish minutes) is fine. If it's O(hours), I'm guessing that's bad but that it might already be happening occasionally and we don't have visibility into it? Are there metrics for we could/should enable/track to learn how long aggressive hard delete transactions have to wait before the immutable timestamp passes?
2. Regarding defining a requirement, we should probably start an email chain with the relevant people.
3. There shouldn't be a distinction between how quickly we need to delete the value vs the row/column keys. From the customer's perspective, it's the same problem: "the data is still in the DB." There is a distinction right now, which motivated me bringing up the scrubber work.

Agreed that it'd be useful to have metrics about how behind targeted sweep gets at some representative places that use the big internal product before deciding to reimplement scrub using the targeted sweep queue versus its own queue.

Agreed that violating the transaction protocol sounds scary.

If targeted sweep gets us to a world where sweep is always caught up, whether the data swept away is from a hard delete or not, then perhaps a scrubber wouldn't even be needed. But my guess is that wouldn't be the case for every deployment at all points in time. Having a separate, higher priority queue for just hard deleted data helps alleviate some of that risk. It also makes it easy for deployments to quickly query whether all hard deleted data has been removed (i.e. the hard delete CLI has a --waitForScrub option that blocks until scrub has finished its work), and/or to devote more resources to the problem with more precision (num scrub threads).

**Comment by j-baker** [comment]
Our experience has been that a single targeted sweep thread can handle write rates of up to 2.5 million writes per hour, but it shards. The gist is that by increasing the number of shards you can arbitrarily weight it so that cleanup is fast enough.

Internally, I can show you examples of consistent 500 writes per second where targeted sweep never falls behind the immutable timestamp by more than 15 seconds according to metrics.

**Comment by j-baker** [comment]
that said, it's likely we'd need to add another targeted sweep queue, at least for conservative hard-delete txns.

One thing I'd like to ask is whether we could plausibly get rid of the notion of hard delete transactions as a whole from AtlasDB. We can discuss this internally, but with targeted sweep we've been able to guarantee that all deletes are cleaned up in a reasonable amount of time.

For example, would we meet our requirements if we could guarantee that hard deleted things were inaccessible within a day? Targeted sweep can absolutely guarantee that in a way that's not possible with hard delete transactions.

So, it'd be good to list here the customer-facing criteria for hard delete transactions (we can do this internally), if only to work out if a reasonable end state is to kill the functionality.
